### PR TITLE
ENH: Enable activation checkpointing for Resnet and SwinTransformer

### PR DESCRIPTION
**Pull request by kenza-bouzid**
Add custom forward that checkpoints activations to trade compute for memory

### CODE REVIEW

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 1]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
+        self,
+        tile_size: int,
+        n_channels: int = 3,
+        checkpoint_activations: bool = False,
+        checkpoint_segments_size: int = 2,
+        bn_momentum: Optional[float] = None,
+    ) -> None:
+        """
+        :param tile_size: The size of the input tiles.
+        :param n_channels: The number of channels in the input tiles.
```
Comment (harshita-s) [inline hi-ml-cpath/src/health_cpath/models/encoders.py:199]: please mention default values, same for other arguments in the doc

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 2]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
```
Comment (harshita-s) [inline hi-ml-cpath/src/health_cpath/models/encoders.py:231]: please mention default argument values, same for other arguments

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 3]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
+    def _custom_forward(self, images: torch.Tensor) -> torch.Tensor:
+        """Custom forward pass that uses activation checkpointing to save memory."""
+        segments = self.checkpoint_segments_size
+        first_layers = [
+            self.feature_extractor_fn.conv1,
+            self.feature_extractor_fn.bn1,
+            self.feature_extractor_fn.relu,
+            self.feature_extractor_fn.maxpool,
+        ]
+        images = checkpoint_sequential(first_layers, segments, images)
```
Comment (harshita-s) [inline hi-ml-cpath/src/health_cpath/models/encoders.py:174]: could we briefly mention difference between `checkpoint_sequential` and `checkpoint` in the comments below

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 4]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
+            self,
+            resnet18,
+            tile_size,
+            n_channels,
+            apply_imagenet_preprocessing=False,
+            checkpoint_activations=checkpoint_activations,
+        )
+        ResNetCheckpointingMixin.__init__(
+            self, self.feature_extractor_fn, bn_momentum, checkpoint_segments_size, checkpoint_activations
+        )
```
Comment (harshita-s) [inline hi-ml-cpath/src/health_cpath/models/encoders.py:247]: same comment about positional/named arguments

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 5]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
+        n_channels: int = 3,
+        checkpoint_activations: bool = False,
+        checkpoint_segments_size: int = 2,
+        bn_momentum: Optional[float] = None,
+    ) -> None:
+        """
+        :param tile_size: The size of the input tiles.
+        :param n_channels: The number of channels in the input tiles.
+        :param checkpoint_activations: Whether to checkpoint activations during forward pass.
+        :param checkpoint_segments_size: The size of checkpointed segments in sequential layers.
```
Comment (harshita-s) [inline hi-ml-cpath/src/health_cpath/models/encoders.py:263]: same comments as above about default argument values

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 6]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
+        tile_size: int,
+        n_channels: int = 3,
+        checkpoint_activations: bool = False,
+        checkpoint_segments_size: int = 2,
+        bn_momentum: Optional[float] = None,
+    ) -> None:
+        """
+        :param tile_size: The size of the input tiles.
+        :param n_channels: The number of channels in the input tiles.
+        :param checkpoint_activations: Whether to checkpoint activations during forward pass.
```
Comment (harshita-s) [inline hi-ml-cpath/src/health_cpath/models/encoders.py:292]: default values please

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 7]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
+            checkpoint_activations is True. If None, sqrt of the default momentum is used.
+        """
+        ImageNetEncoder.__init__(
+            self,
+            resnet50,
+            tile_size,
+            n_channels,
+            apply_imagenet_preprocessing=False,
+            checkpoint_activations=checkpoint_activations,
+        )
```
Comment (harshita-s) [inline hi-ml-cpath/src/health_cpath/models/encoders.py:304]: same as above for positional/named

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 8]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
+        """
+        ImageNetEncoder.__init__(
+            self,
+            swin_tiny_patch4_window7_224,
+            tile_size,
+            n_channels,
+            apply_imagenet_preprocessing=False,
+            checkpoint_activations=checkpoint_activations,
+        )
+        SwinTransformerCheckpointingMixin.__init__(self, self.feature_extractor_fn, checkpoint_segments_size)
```
Comment (harshita-s) [inline hi-ml-cpath/src/health_cpath/models/encoders.py:369]: same comments as above

File: hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py [hunk 9]
```diff
@@ -65,6 +65,18 @@ class EncoderParams(param.Parameterized):
     projection_dim: int = param.Integer(
         default=0, doc="If > 0, project the encoded tiles to this dimension. Otherwise, use identity projection."
     )
+    checkpoint_encoder: bool = param.Boolean(
+        default=False, doc="If True, checkpoint the encoder gradients during training. This is useful to trade compute "
+        "for memory, and is only supported for the Resnet18 and Resnet50 encoders at the moment. Default: False."
```
Comment (harshita-s) [inline hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py:70]: same comment as above about the docs of `checkpoint_activations` can we keep the same name for the container parameter?

File: hi-ml-cpath/testhisto/testhisto/models/test_deepmil.py [hunk 10]
```diff
@@ -720,3 +731,81 @@ def test_ssl_containers_default_checkpoint(container_type: BaseMIL) -> None:
...
+    if encoder_type != SwinTransformer_NoPreproc.__name__:  # SwinT requires images of 224 input size, mock tiles are 28
+        # 2. Train the model with and without checkpointing and compare the encoder parameters
+        trainer_no_ckpt = Trainer(max_epochs=1, limit_train_batches=2, limit_val_batches=2, limit_test_batches=2)
+        trainer_no_ckpt.fit(model_no_ckpt_enc, train_dataloader, val_dataloader)
+
+        trainer_ckpt = Trainer(max_epochs=1, limit_train_batches=2, limit_val_batches=2, limit_test_batches=2)
+        trainer_ckpt.fit(model_ckpt_enc, train_dataloader, val_dataloader)
+
+        for param_no_ckpt, param in zip(model_no_ckpt_enc.encoder.parameters(), model_ckpt_enc.encoder.parameters()):
+            assert torch.allclose(param_no_ckpt, param, atol=1e-4)
```
Comment (harshita-s) [inline hi-ml-cpath/testhisto/testhisto/models/test_deepmil.py:798]: can we parametrize bn_momentum? do we expect to see this with different bn_momentum values?

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 11]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
+
+        # Set momentum for the first batch norm layer
+        self.feature_extractor_fn.bn1.momentum = _momentum
+
+        def _set_bn_momentum(layer_block: nn.Sequential) -> None:
+            for sub_layer in layer_block:
+                for key, layer in sub_layer._modules.items():
+                    if "bn" in key:
+                        assert isinstance(layer, nn.BatchNorm2d), "Expected BatchNorm2d layer."
+                        layer.momentum = _momentum
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/models/encoders.py:157]: Maybe overcautious, but would be safer to loop over the layers and modify only those where `isinstance(layer, nn.Batchnorm2d)`? ie, not relying on the naming scheme?

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 12]
```diff
@@ -2,34 +2,42 @@
...
+        self,
+        tile_size: int = 0,
+        n_channels: int = 3,
+        input_dim: Optional[Sequence[int]] = None,
+        checkpoint_activations: bool = False,
+    ) -> None:
         """The `TileEncoder` constructor should be called after setting any attributes needed in
         `_get_preprocessing()` or `_get_encoder()`.
 
         :param tile_size: Tile width/height, in pixels.
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:36]: ... inparticular since you are making tile_size a required argument elsewhere

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 13]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
+        images = checkpoint_sequential(self.feature_extractor_fn.layer3, segments, images)
+        images = checkpoint_sequential(self.feature_extractor_fn.layer4, segments, images)
+
+        images = checkpoint(self.feature_extractor_fn.avgpool, images)
+        images = torch.flatten(images, 1)
+        images = checkpoint(self.feature_extractor_fn.fc, images)
+        return images
+
+
+class Resnet18(ResNetCheckpointingMixin, ImageNetEncoder):
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/models/encoders.py:186]: Maybe choose a different class name? `Resnet18withCheckpointing`?

File: hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py [hunk 14]
```diff
@@ -65,6 +65,18 @@ class EncoderParams(param.Parameterized):
     projection_dim: int = param.Integer(
         default=0, doc="If > 0, project the encoded tiles to this dimension. Otherwise, use identity projection."
     )
+    checkpoint_encoder: bool = param.Boolean(
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py:68]: `use_encoder_checkpointing`?

File: hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py (hunk 9 above)
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py:70]: Resnet and SwinT?

File: hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py [hunk 15]
```diff
@@ -65,6 +65,18 @@ class EncoderParams(param.Parameterized):
...
+        default=False, doc="If True, checkpoint the encoder gradients during training. This is useful to trade compute "
+        "for memory, and is only supported for the Resnet18 and Resnet50 encoders at the moment. Default: False."
+    )
+    checkpoint_segments_size: int = param.Integer(
+        default=2,
+        bounds=(1, None),
+        doc="The segments size to use for checkpointing the encoder's activations. Default: 2. "
+    )
```
Comment (ant0nsc) [inline hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py:76]: segments_size -> segment_size?

File: hi-ml-cpath/src/health_cpath/models/encoders.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:36]: Good point, I have no idea tbh, It used to be 0, should we maybe set it to 224?

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 16]
```diff
@@ -2,34 +2,42 @@
...
+        checkpoint_activations: bool = False,
+    ) -> None:
         """The `TileEncoder` constructor should be called after setting any attributes needed in
         `_get_preprocessing()` or `_get_encoder()`.
 
         :param tile_size: Tile width/height, in pixels.
         :param n_channels: Number of channels in the tile (default=3).
         :param input_dim: Input shape, to override default of `(n_channels, tile_size, tile_size)`.
+        :param checkpoint_activations: Whether to checkpoint activations during forward pass. This can be used to trade
+            compute for memory. Note that this is only supported for some encoders.
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:40]: Resnets and SwinTranformer at the moment, I can clarify it in the docs

File: hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py (hunk 9 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py:70]: yep, forgot to update the docs after adding SwinT, thanks!

File: hi-ml-cpath/src/health_cpath/models/encoders.py (hunk 13 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:186]: Oh I wouldn't change it as it can be used with and without checkpointing, we can enable checkpointning from the container via encoder params checkpoint_activations (I'll rename it as you suggest ;) )

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 17]
```diff
@@ -2,34 +2,42 @@
...
+        input_dim: Optional[Sequence[int]] = None,
+        checkpoint_activations: bool = False,
+    ) -> None:
         """The `TileEncoder` constructor should be called after setting any attributes needed in
         `_get_preprocessing()` or `_get_encoder()`.
 
         :param tile_size: Tile width/height, in pixels.
         :param n_channels: Number of channels in the tile (default=3).
         :param input_dim: Input shape, to override default of `(n_channels, tile_size, tile_size)`.
+        :param checkpoint_activations: Whether to checkpoint activations during forward pass. This can be used to trade
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:39]: done thanks!

File: hi-ml-cpath/src/health_cpath/models/encoders.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:40]: done, thanks

[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:40]: done!

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 18]
```diff
@@ -62,18 +77,26 @@ def _get_encoder(self) -> Tuple[Callable, int]:
...
+        apply_imagenet_preprocessing: bool = True,
+        checkpoint_activations: bool = False,
+    ) -> None:
         """
         :param feature_extraction_model: A function accepting a `pretrained` keyword argument that
         returns a classifier pretrained on ImageNet, such as the ones from `torchvision.models.*`.
         :param tile_size: Tile width/height, in pixels.
         :param n_channels: Number of channels in the tile (default=3).
         :param apply_imagenet_preprocessing: Whether to apply ImageNet preprocessing to the input.
+        :param checkpoint_activations: Whether to checkpoint activations during forward pass. This can be used to trade
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:94]: done

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 19]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
-class Resnet18(ImageNetEncoder):
-    def __init__(self, tile_size: int, n_channels: int = 3) -> None:
-        super().__init__(resnet18, tile_size, n_channels, apply_imagenet_preprocessing=True)
+class ResNetCheckpointingMixin:
+    """Mixin class for checkpointing activations in ResNet-based encoders."""
+
+    def __init__(
+        self,
+        feature_extractor_fn: ResNet,
+        bn_momentum: Optional[float] = None,
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:118]: done

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 20]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
+        self.feature_extractor_fn = feature_extractor_fn
+        self.checkpoint_segments_size = checkpoint_segments_size
+        self.bn_momentum = bn_momentum
+        if checkpoint_activations:
+            self._set_batch_norm_momentum()
+
+    def _set_batch_norm_momentum(self) -> None:
+        """Set the momentum of batch norm layers in the ResNet model to avoid running statistics from going out of
+        sync due to activations checkpointing. The forward pass is applied twice which results in double updates of
+        these statistics. We can workaround that by using sqrt of default momentum.
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:141]: done!

File: hi-ml-cpath/src/health_cpath/models/encoders.py (hunk 11 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:157]: good point, done!

File: hi-ml-cpath/src/health_cpath/models/encoders.py (hunk 3 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:174]: done

File: hi-ml-cpath/src/health_cpath/models/encoders.py [hunk 21]
```diff
@@ -86,32 +109,264 @@ def _get_encoder(self) -> Tuple[torch.nn.Module, int]:
...
+            self,
+            resnet18,
+            tile_size,
+            n_channels,
+            apply_imagenet_preprocessing=True,
+            checkpoint_activations=checkpoint_activations,
+        )
+        ResNetCheckpointingMixin.__init__(
+            self, self.feature_extractor_fn, bn_momentum, checkpoint_segments_size, checkpoint_activations
+        )
```
[the remark this answers is not in the document]
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/models/encoders.py:215]: done

File: hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py (hunk 15 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/src/health_cpath/utils/deepmil_utils.py:76]: pytorch uses segments

File: hi-ml-cpath/testhisto/testhisto/models/test_deepmil.py (hunk 10 above)
Comment (kenza-bouzid) [reply hi-ml-cpath/testhisto/testhisto/models/test_deepmil.py:798]: good point, added it thanks!

**Review by harshita-s — APPROVED** [review]
